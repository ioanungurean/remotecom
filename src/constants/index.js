export const HOST = 'http://localhost:4002';

export const PEOPLE_COLUMNS = [
  { label: 'Name', id: 'name' },
  { label: 'Role', id: 'role' },
  { label: 'Type', id: 'type' },
  { label: 'Country', id: 'country' },
  { label: 'Salary', id: 'salary', align: 'right' },
  { label: '', id: 'actions', align: 'right' },
];

export const EMPLOYMENT_TYPE = [
  { label: 'Employee', id: 'employee' },
  { label: 'Contractor', id: 'contractor' },
];
