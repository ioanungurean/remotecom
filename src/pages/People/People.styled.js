import styled from 'styled-components';

import { ReactComponent as IconUser } from 'theme/icons/user.svg';
import LoadingLogo from 'components/LoadingLogo';
import Text, { TextLight } from 'components/Text';

export const Container = styled.main`
  margin: 42px auto;
  width: 100%;
  max-width: var(--layout-width);
`;

export const Wrapper = styled.div`
  display: flex;
  margin-bottom: 32px;
  justify-content: space-between;
  align-items: center;
`;

export const Header = styled.div`
  display: flex;
  align-items: center;

  & > :not(:last-child) {
    margin: 0 8px 6px 0;
  }
`;

export const StyledIconUser = styled(IconUser)`
  fill: var(--colors-blank);
  margin-right: 10px;
`;

export const StyledLoadingLogo = styled(LoadingLogo)`
  display: flex;
  margin: 80px auto auto auto;
`;

export const StyledTextLight = styled(TextLight)`
  ${({ theme }) => theme.typography.bodyCaption};
`;

export const Filters = styled.div`
  display: flex;
  margin-bottom: 16px;

  justify-content: space-between;
`;

export const RightSide = styled.div`
  display: flex;

  & > :not(:last-child) {
    margin-right: 16px;
  }
`;

export const NotFound = styled(Text)`
  display: flex;
  padding: 18px 0;
`;
