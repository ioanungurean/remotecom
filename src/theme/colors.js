import { css } from 'styled-components';

const colors = css`
  --colors-irisBlue: #624de3;
  --colors-royalPurple: #6447ff;
  --colors-darkBlue: #00234b;
  --colors-redPink: #ff4a5a;
  --colors-zircon: #fdfeff;
  --colors-selago: #e7effc;
  --colors-cosmos: #ffdbde;

  --colors-greekLilac: #b7b8eb;
  --colors-oldLavander: #d0caf7;

  --colors-bayoux: #525f7f;
  --colors-lynch: #617798;
  --colors-pigeon: #a8bdd4;
  --colors-heather: #b1becd;
  --colors-coldIceGrey: #c6d6ef;
  --colors-mischka: #e8ecee;
  --colors-catskillWhite: #eaf0f6;
  --colors-linkWater: #f4f7fc;

  --colors-blank: #ffffff;
`;

export default colors;
