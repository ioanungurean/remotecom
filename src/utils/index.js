export const fomatCurrency = (amount, currency) => {
  const formattedCurrency = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency,
    currencyDisplay: 'code',
    minimumFractionDigits: 2,
  }).format(amount);

  const symbol = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency,
    currencyDisplay: 'symbol',
    minimumFractionDigits: 2,
  })
    .formatToParts(amount)
    .filter(({ type }) => type === 'currency')
    .map(({ value }) => value)
    .toString();

  return `${symbol} ${formattedCurrency}`;
};
