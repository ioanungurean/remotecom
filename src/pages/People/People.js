import { useState } from 'react';
import { Link, useHistory } from 'react-router-dom';
import { useQuery } from 'react-query';
import { useDebounce } from 'use-debounce';

import { fetchPeople } from 'api/people';
import { fomatCurrency } from 'utils/index';
import { PEOPLE_COLUMNS, EMPLOYMENT_TYPE } from 'constants/index';
import Button from 'components/Button';
import Searchbox from 'components/Searchbox';
import Checkbox from 'components/Checkbox';
import { Card, CardBody } from 'components/Card';
import {
  TableWrapper,
  Table,
  TableHead,
  TableBody,
  TableRow,
  TableThCell,
  TableCell,
} from 'components/Table';
import Text from 'components/Text';

import {
  Container,
  Wrapper,
  Header,
  StyledIconUser,
  StyledTextLight,
  StyledLoadingLogo,
  Filters,
  RightSide,
  NotFound,
} from './People.styled';

export default function People() {
  const history = useHistory();
  const [search, setSearch] = useState('');
  const [debuncedSearch] = useDebounce(search, 300);
  const [employmentTypes, setEmploymentTypes] = useState([]);
  const { isLoading, data } = useQuery(['fetchPeople', employmentTypes, debuncedSearch], () =>
    fetchPeople(employmentTypes, debuncedSearch)
  );

  const handleAddNewMemberClick = () => {
    history.push('/people/new');
  };

  const handlEmploymentChange = (newEmployment) => {
    if (employmentTypes.includes(newEmployment)) {
      setEmploymentTypes(employmentTypes.filter((type) => type !== newEmployment));
    } else {
      setEmploymentTypes(employmentTypes.concat(newEmployment));
    }
  };

  const renderRows = ({ id, name, jobTitle, employment, country, salary, currency }) => {
    return (
      <TableRow key={id}>
        <TableCell emphasize>{name}</TableCell>
        <TableCell>{jobTitle}</TableCell>
        <TableCell>{employment}</TableCell>
        <TableCell>{country}</TableCell>
        <TableCell align="right">{fomatCurrency(salary, currency)}</TableCell>
        <TableCell align="right">
          <Link to={`/people/edit/${id}`}>Edit</Link>
        </TableCell>
      </TableRow>
    );
  };

  return (
    <Container>
      <Wrapper>
        <Header>
          <Text size="h2">People</Text>
          {data && data.length > 0 && <StyledTextLight>{data.length} employees</StyledTextLight>}
        </Header>
        <Button onClick={handleAddNewMemberClick}>
          <StyledIconUser title="Add a new member" />
          Add member
        </Button>
      </Wrapper>
      <Card>
        <CardBody>
          <Filters>
            <Searchbox value={search} placeholder="Search employees..." onChange={setSearch} />
            <RightSide>
              {EMPLOYMENT_TYPE.map(({ id, label }) => (
                <Checkbox key={id} id={id} label={label} onChange={handlEmploymentChange} />
              ))}
            </RightSide>
          </Filters>
          <TableWrapper>
            <Table>
              <TableHead>
                <TableRow>
                  {PEOPLE_COLUMNS.map(({ label, id, align }) => (
                    <TableThCell key={id} align={align}>
                      {label}
                    </TableThCell>
                  ))}
                </TableRow>
              </TableHead>
              <TableBody>
                {isLoading && (
                  <tr>
                    <td colSpan={PEOPLE_COLUMNS.length}>
                      <StyledLoadingLogo />
                    </td>
                  </tr>
                )}
                {data && data.length > 0 && <>{data.map(renderRows)}</>}
                {data && data.length === 0 && (
                  <tr>
                    <td>
                      <NotFound size="bodySmallBold">No employees found...</NotFound>
                    </td>
                  </tr>
                )}
              </TableBody>
            </Table>
          </TableWrapper>
        </CardBody>
      </Card>
      <Link to="/playground">Playground</Link>
    </Container>
  );
}
