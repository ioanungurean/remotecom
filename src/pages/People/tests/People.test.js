import { render, screen, fireEvent, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { QueryClient, QueryClientProvider } from 'react-query';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

import { HOST } from '../../../constants/index';
import { fomatCurrency } from '../../../utils/index';
import { theme } from '../../../theme';
import People from '../index';
import { allPeople, peopleFilteredByName, peopleFilteredByEmployment } from '../../../mocks';

const server = setupServer(
  rest.get(`${HOST}/people`, (req, res, ctx) => {
    const query = req.url.searchParams;
    const name_like = query.get('name_like');
    const employment = query.get('employment');

    if (name_like) {
      if (name_like === 'no') {
        return res(ctx.status(200), ctx.json([]));
      }

      return res(ctx.status(200), ctx.json(peopleFilteredByName));
    }

    if (employment) {
      return res(ctx.status(200), ctx.json(peopleFilteredByEmployment));
    }

    return res(ctx.status(200), ctx.json(allPeople));
  })
);

const renderComponent = () => {
  const queryClient = new QueryClient();

  return render(
    <MemoryRouter>
      <ThemeProvider theme={theme}>
        <QueryClientProvider client={queryClient}>
          <People />
        </QueryClientProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('People', () => {
  it('renders correctly', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render the table with the correct data', async () => {
    renderComponent();
    await waitFor(() => expect(screen.getAllByRole('row')).toHaveLength(allPeople.length + 1)); // +1 for the table header row
  });
  it('should change the page when Add member button is clicked', () => {
    renderComponent();
    fireEvent.click(screen.getByRole('button', { name: /Add member/i }));
    expect(screen.getByText(/Add member/i)).toBeInTheDocument();
  });

  it('should render a row correctly', async () => {
    renderComponent();
    await screen.findByText(allPeople[1].name);

    const rows = await screen.findAllByRole('row');

    expect(within(rows[1]).getByText(allPeople[0].name)).toBeInTheDocument();
    expect(within(rows[1]).getByText(allPeople[0].jobTitle)).toBeInTheDocument();
    expect(within(rows[1]).getByText(allPeople[0].employment)).toBeInTheDocument();
    expect(within(rows[1]).getByText(allPeople[0].country)).toBeInTheDocument();
    expect(
      within(rows[1]).getByText(
        fomatCurrency(allPeople[0].salary, allPeople[0].currency).replace(/\xa0/g, ' ')
      )
    ).toBeInTheDocument();
  });

  it('should update the table when searching', async () => {
    renderComponent();
    userEvent.type(screen.getByRole('textbox'), 'Something');
    expect(screen.getByRole('textbox')).toHaveValue('Something');

    await waitFor(
      () => expect(screen.getAllByRole('row')).toHaveLength(peopleFilteredByName.length + 1) // +1 for the table header row
    );

    expect(screen.getByText(peopleFilteredByName[0].name)).toBeInTheDocument();
    expect(screen.getByText(peopleFilteredByName[0].name)).toBeInTheDocument();
  });

  it('should show proper message when there are no results', async () => {
    renderComponent();
    userEvent.type(screen.getByRole('textbox'), 'no');
    expect(screen.getByRole('textbox')).toHaveValue('no');

    await waitFor(() => expect(screen.getByText('No employees found...')).toBeInTheDocument());
  });

  it('should update the table when a checkbox is clicked', async () => {
    renderComponent();

    userEvent.click(screen.getByText('Employee'));
    expect(screen.getByLabelText('Employee')).toBeChecked();

    await waitFor(
      () => expect(screen.getAllByRole('row')).toHaveLength(peopleFilteredByName.length + 1) // +1 for the table header row
    );

    expect(screen.getByText(peopleFilteredByEmployment[0].name)).toBeInTheDocument();
    expect(screen.getByText(peopleFilteredByEmployment[0].name)).toBeInTheDocument();
  });

  it('the Edit links should have the correct href', async () => {
    renderComponent();

    const editLinks = await screen.findAllByText('Edit');

    expect(editLinks[0]).toHaveAttribute('href', `/people/edit/${allPeople[0].id}`);
    expect(editLinks[3]).toHaveAttribute('href', `/people/edit/${allPeople[3].id}`);
  });
});
