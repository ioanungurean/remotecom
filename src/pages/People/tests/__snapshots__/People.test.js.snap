// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`People renders correctly 1`] = `
<div>
  <main
    class="sc-dmlrTW kRPEfO"
  >
    <div
      class="sc-kfzAmx Crazx"
    >
      <div
        class="sc-fKFyDc fkrfJR"
      >
        <span
          class="sc-fFubgz dcpucn"
        >
          People
        </span>
      </div>
      <button
        class="sc-bdfBwQ dxIjpe"
        type="button"
      >
        <svg
          class="sc-bBXqnf dYxpNr"
          title="Add a new member"
        >
          user.svg
        </svg>
        Add member
      </button>
    </div>
    <div
      class="sc-fubCfw foltsl"
    >
      <div
        class="sc-jrAGrp iRSOhu"
      >
        <div
          class="sc-lmoMRL krFKNZ"
        >
          <label
            class="sc-dlfnbm cQtXG"
            for="search"
          >
            <svg
              class="sc-hKgILt eTouyH"
              title="Search employees"
            >
              search.svg
            </svg>
            <input
              class="sc-gsTCUz cdNqGK"
              id="search"
              placeholder="Search employees..."
              value=""
            />
          </label>
          <div
            class="sc-iJuUWI OOrbr"
          >
            <label
              class="sc-gKsewC RSzGc"
              for="employee"
            >
              <input
                class="sc-jSgupP lcELIj"
                id="employee"
                type="checkbox"
              />
              <span
                class="sc-eCssSg gKrttk"
              />
              <span
                class="sc-iBPRYJ iyZyAb"
              >
                Employee
              </span>
            </label>
            <label
              class="sc-gKsewC RSzGc"
              for="contractor"
            >
              <input
                class="sc-jSgupP lcELIj"
                id="contractor"
                type="checkbox"
              />
              <span
                class="sc-eCssSg gKrttk"
              />
              <span
                class="sc-iBPRYJ iyZyAb"
              >
                Contractor
              </span>
            </label>
          </div>
        </div>
        <div
          class="sc-iqHYGH fxwATU"
        >
          <table
            class="sc-crrsfI fxfLfE"
          >
            <thead
              class="sc-dQppl"
            >
              <tr
                class="sc-kstrdz ixCVwl"
              >
                <th
                  class="sc-hBEYos inVeRO"
                >
                  Name
                </th>
                <th
                  class="sc-hBEYos inVeRO"
                >
                  Role
                </th>
                <th
                  class="sc-hBEYos inVeRO"
                >
                  Type
                </th>
                <th
                  class="sc-hBEYos inVeRO"
                >
                  Country
                </th>
                <th
                  class="sc-hBEYos fHqFnR"
                >
                  Salary
                </th>
                <th
                  class="sc-hBEYos fHqFnR"
                />
              </tr>
            </thead>
            <tbody
              class="sc-bqyKva"
            >
              <tr>
                <td
                  colspan="6"
                >
                  <div
                    class="sc-hHftDr bTJjW sc-iwyYcG iyoACc"
                  >
                    <svg
                      version="1.1"
                      viewBox="0 0 119.9 148.4"
                      x="0px"
                      xml:space="preserve"
                      xmlns="http://www.w3.org/2000/svg"
                      y="0px"
                    >
                      <defs />
                      <g>
                        <g>
                          <path
                            class="sc-idOhPF"
                            d="M49.1,117.1C41.5,112.7,35.4,106.6,31,99c-4.4-7.6-6.7-15.9-6.7-24.8c0-9.1,2.2-17.4,6.7-25.1
 c4.4-7.6,10.5-13.7,18.1-18.1c7.6-4.4,15.9-6.6,24.9-6.7V0C60.7,0,48.3,3.3,37,9.9C25.5,16.5,16.5,25.5,9.9,37
 C3.3,48.4,0,60.8,0,74.2c0,13.4,3.3,25.8,9.9,37.1c6.6,11.3,15.6,20.3,27.1,27c11.4,6.7,23.7,10,37.1,10.1v-24.6
 C65.1,123.7,56.8,121.5,49.1,117.1z"
                          />
                        </g>
                        <g>
                          <path
                            class="sc-dIUggk"
                            d="M74.1,28.5V120c25.3,0,45.8-20.5,45.8-45.8S99.4,28.5,74.1,28.5z"
                          />
                        </g>
                      </g>
                    </svg>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <a
      href="/playground"
    >
      Playground
    </a>
  </main>
</div>
`;
